test/test_helpers.jl

```diff
@@ -1,3 +1,100 @@
+using JLArrays  # provides a fake GPU array
+JLArrays.allowscalar(false)
+
+using Adapt
+jl32(xs) = adapt(JLArray{Float32}, xs)
+Adapt.adapt_storage(::Type{<:JLArray{Float32}}, xs::AbstractArray{<:Complex}) = convert(JLArray{ComplexF32}, xs)
+Adapt.adapt_storage(::Type{<:JLArray{Float32}}, x::Float64) = Float32(x)
+Adapt.adapt_storage(::Type{<:JLArray{Float32}}, x::ComplexF64) = ComplexF32(x)
+
+f32(xs) = adapt(Array{Float32}, xs)  # these adapt methods are piracy, just for testing though!
+Adapt.adapt_storage(::Type{<:Array{Float32}}, xs::AbstractArray{<:Complex}) = convert(Array{ComplexF32}, xs)
+Adapt.adapt_storage(::Type{<:Array{Float32}}, x::Float64) = Float32(x)
+Adapt.adapt_storage(::Type{<:Array{Float32}}, x::ComplexF64) = ComplexF32(x)
+
+Adapt.adapt(T::Type{<:JLArray{Float32}}, x::AbstractThunk) = adapt(T, unthunk(x))
+Adapt.adapt(T::Type{<:Array{Float32}}, x::AbstractThunk) = adapt(T, unthunk(x))
```

Comment:
Can you explain the purpose of this stuff in a comment?

Reply:
Done.